docs/internal-documentation/maintenance/Garbage-Collection-Testing.md

```diff
@@ -0,0 +1,122 @@
+The purpose of this instruction is to serve as a notice for manual testing of the new garbage collection feature at release time.
+
+## Testing the unfinished tests garbage collection feature
+
+1. Ensure that the database has the required aditionnal columns for this feature:
+```
+SELECT nb_retries FROM test_results LIMIT 0;
+```
+Should return:
+
+```
+ nb_retries 
+------------
+(0 rows)
+```
+
+2. Check that your backend_config.ini has the proper parameter set
+
+Either disabled:
+```
+#maximal_number_of_retries=3
+```
+or set to 0:
+```
+maximal_number_of_retries=0
+```
+
+3. Start a test and wait for it to be finished
+
+```
+SELECT hash_id, progress FROM test_results LIMIT 1;
+```
+Should return:
+
+```
+     hash_id      | progress 
+------------------+----------
+ 3f7a604683efaf93 |      100
+(1 row)
+```
+
+4. Sumulate a crashed test
+```
+UPDATE test_results SET progress = 50, test_start_time = '2020-01-01' WHERE hash_id = '3f7a604683efaf93';
+```
+
+5. Chaek that the backend finishes the test with a result stating it was unfinished
+
+```
+SELECT hash_id, progress FROM test_results WHERE hash_id = '3f7a604683efaf93';
+```
+Should return a finished result:
+```
+     hash_id      | progress 
+------------------+----------
+ 3f7a604683efaf93 |      100
+(1 row)
+```
+6. Ensure the test result contains the backend generated critical message message:
+```
+{"tag":"UNABLE_TO_FINISH_TEST","level":"CRITICAL","timestamp":"300","module":"BACKEND_TEST_AGENT"}
+```
+
+```
+SELECT hash_id, progress FROM test_results WHERE hash_id = '3f7a604683efaf93' AND results::text like '%UNABLE_TO_FINISH_TEST%';
+```
+Should return:
+```
+     hash_id      | progress 
+------------------+----------
+ 3f7a604683efaf93 |      100
+(1 row)
+
+```
+
+## Testing the unfinished tests garbage collection feature with a number of retries set to allow finishing tests without a critical error
+
+1. Set the maximal_number_of_retries parameter to a value greater than 0 in the backend_config.ini config file 
+```
+maximal_number_of_retries=1
+```
+
+2. Restart the test agent in order for the parameter to be taken into account
+```
+zonemaster_backend_testagent restart
+```
+
+3. Sumulate a crashed test
+```
+UPDATE test_results SET progress = 50, test_start_time = '2020-01-01' WHERE hash_id = '3f7a604683efaf93';
+```
+
+4. Check that the backend finishes the test WITHOUT a result stating it was unfinished
+
+```
+SELECT hash_id, progress FROM test_results WHERE hash_id = '3f7a604683efaf93';
+```
+Should return a finished result:
+```
+     hash_id      | progress 
+------------------+----------
+ 3f7a604683efaf93 |      100
+(1 row)
+```
+5. Ensure the test result does NOT contain the backend generated critical message message:
+```
+{"tag":"UNABLE_TO_FINISH_TEST","level":"CRITICAL","timestamp":"300","module":"BACKEND_TEST_AGENT"}
+```
+
+```
+SELECT hash_id, progress FROM test_results WHERE hash_id = '3f7a604683efaf93' AND results::text like '%UNABLE_TO_FINISH_TEST%';
```

Comment:
Have you considered using `not like` instead of `like` here?

Reply:
Mathias I think keeping the like here is the right option. What we want is to ensure the 
```
{"tag":"UNABLE_TO_FINISH_TEST","level":"CRITICAL","timestamp":"300","module":"BACKEND_TEST_AGENT"}
```
message is not present in the database and thus the SQL query returns:
```
hash_id | progress 
---------+----------
(0 rows)
```

If I use NOT LIKE here and the database contains other (unrelated tests) they are going to be returned. It seems to much much easier to check there is no result.